### PR TITLE
Add types to package sse4_crc32

### PR DESCRIPTION
**Pull request by b8s256**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/sse4_crc32/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+export function fromStream(stream: ReadStream, crc?: number): Crc32CStream;
+/**
+ *
+ * @param buf Buffer object or string
+ * @param initial Non-zero integer value
+ */
+export function calculate(buf: Buffer | string, initial?: number): number;
+
+// only available when NODE_ENV='production'
+export const hardware_support: boolean;
```
Comment (peterblazejewicz) [inline types/sse4_crc32/index.d.ts:29]: what about?
```suggestion
export const hardware_support?: boolean | undefined;
```
so people can enforce this explicitly when debugging/benchmarking
```ts
mod.hardware_support!;
```

File: types/sse4_crc32/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import { Writable } from 'stream';
+
+export class Crc32CStream extends Writable {
+    constructor(crc?: number);
+}
+
+/**
+ *
+ * @param stream Readable stream object
+ * @param crc Non-zero integer value
```
Comment (peterblazejewicz) [inline types/sse4_crc32/index.d.ts:18]:
```suggestion
 * @param [crc] Non-zero integer value
```

File: types/sse4_crc32/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ */
+export function fromStream(stream: ReadStream, crc?: number): Crc32CStream;
+/**
+ *
+ * @param buf Buffer object or string
+ * @param initial Non-zero integer value
```
Comment (peterblazejewicz) [inline types/sse4_crc32/index.d.ts:24]:
```suggestion
 * @param [initial] Non-zero integer value
```

File: types/sse4_crc32/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+export function calculate(buf: Buffer | string, initial?: number): number;
+
+// only available when NODE_ENV='production'
+export const hardware_support: boolean | undefined;
+/**
+ * Calculates CRC-32C for the specified string/buffer using SSE 4.2 extensions
+ * @param buf Buffer object or string
+ * @param [initial] Non-zero integer value
+ */
+export function sse42_crc(buf: Buffer | string, initial?: number): number;
```
Comment (peterblazejewicz) [inline types/sse4_crc32/index.d.ts:35]:
```suggestion
export const  sse42_crc: ((buf: Buffer | string, initial?: number) => number) | undefined;
```

File: types/sse4_crc32/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * Calculates CRC-32C for the specified string/buffer using table-lookups
+ * @param buf Buffer object or string
+ * @param [initial] Non-zero integer value
+ */
+export function table_crc(buf: Buffer | string, initial?: number): number;
```
Comment (peterblazejewicz) [inline types/sse4_crc32/index.d.ts:41]:
```suggestion
export const table_crc: ((buf: Buffer | string, initial?: number) => number) | undefined;
```

File: types/sse4_crc32/sse4_crc32-tests.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import { fromStream, calculate, hardware_support, sse42_crc, table_crc } from 'sse4_crc32';
+
+// $ExpectType Crc32CStream
+fromStream(createReadStream('./test.txt'), 1);
+
+// $ExpectType number
+calculate('123', 1);
+
+// $ExpectType number
+sse42_crc('123', 1);
```
Comment (peterblazejewicz) [inline types/sse4_crc32/sse4_crc32-tests.ts:11]:
```suggestion
sse42_crc!('123', 1);
```

File: types/sse4_crc32/sse4_crc32-tests.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+// $ExpectType number
+table_crc('123', 1);
```
Comment (peterblazejewicz) [inline types/sse4_crc32/sse4_crc32-tests.ts:14]:
```suggestion
table_crc!('123', 1);
```

Comment (peterblazejewicz) [reply types/sse4_crc32/sse4_crc32-tests.ts:14]: this one can be also invoked 'old school':
```ts
table_crc && table_crc('123', 1);
```

File: types/sse4_crc32/index.d.ts (hunk 4 above)
Comment (naeemy) [reply types/sse4_crc32/index.d.ts:35]: ah, I see

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@naeemy thanks!

**Comment by naeemy** [comment]
Ready to merge

**Comment by naeemy** [comment]
Ready to merge

**Pull request merged**